source/common/http/utility.cc

```diff
@@ -380,6 +382,94 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
   }
 }
 
+void Utility::sanitizeConnectionHeader(spdlog::logger& logger, Http::HeaderMap& headers) {
+  // Remove any headers nominated by the Connection header. The TE header
+  // is sanitized and removed only if it's empty after removing unsupported values
+  // See https://github.com/envoyproxy/envoy/issues/8623
+  const auto& cv = Http::Headers::get().ConnectionValues;
+  const auto& connection_header_value = headers.Connection()->value();
+
+  StringUtil::CaseUnorderedSet headers_to_remove{};
+
+  // Split the connection header and evaluate each nominated header
+  for (const auto& token : absl::StrSplit(connection_header_value.getStringView(), ',')) {
+
+    const absl::string_view token_sv = StringUtil::trim(token);
+
+    // If the Connection token value is not a nominated header, ignore it here since
+    // the connection header is removed elsewhere when the H1 request is upgraded to H2
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, cv.Close) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Http2Settings) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.KeepAlive) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Upgrade)) {
+      continue;
+    }
+
+    // Build the LowerCaseString for header lookup
+    const std::string header_to_remove = std::string(token_sv.data(), token_sv.size());
+    const LowerCaseString lcs_header_to_remove(header_to_remove);
+
+    // By default we will remove any nominated headers
+    bool keep_header = false;
+    StringUtil::CaseUnorderedSet tokens_to_remove{};
+
+    // Determine whether the nominated header contains unsupported values
+    HeaderEntry* nominated_header = headers.get(lcs_header_to_remove);
+
+    if (nominated_header) {
+      const HeaderString& nominated_header_value = nominated_header->value();
+      for (const auto& header_value : absl::StrSplit(nominated_header_value.getStringView(), ',')) {
+
+        const absl::string_view header_sv = StringUtil::trim(header_value);
+
+        // Check whether TE contains multiple values and remove everything except "trailers"
+        if (StringUtil::CaseInsensitiveCompare()(token_sv, Http::Headers::get().TE.get()) &&
```

Comment:
This is loop invariant (token_sv == TE) and is required for keep_header so we can specialize this loop for those two cases.

Reply:
Thanks for the feedback. If I have to make more revisions, I will try to incorporate this change.

Reply:
I agree that this loop needs improvement. If I read the intent and standard correctly, if the header is not TE, when it is removed entirely and no splitting of its value is needed. If it is TE header it retained only when the "trailers" keyword is present and the retained form is "TE: trailers", so there is not need for the removeTokens() stuff below, you can just set the TE value to "tokens" if it was present in the original header.
Tokenization is not cheap, it may be a DoS vector if someone creates HTTP request with headers with a lot of tokens and adds the Connection header to remove them.